cisagov/getgov
# Include audit logging

## 🗣 Description ##

The purpose of this PR is to automatically log any model changes to an audit table in the database.

It also includes a refactoring of models.py. To make this work easier to review, changes to the models are in one commit and then moving the models to their new file locations are in a subsequent commit. (Except for a couple of tiny bug fixes while I was getting the tests to run.)

<img src="https://user-images.githubusercontent.com/15718530/201141033-3804b2de-a071-4eca-a619-b07caae903ee.png" width="300px" />

@neilmb I'm tagging you as primary reviewer, but @igorkorenfeld this also touches on your work and I'd welcome any feedback.

## 💭 Motivation and context ##

As implemented, this logs _all_ changes, which includes keeping an audit log of the `last_login` field of the User table. This may not be desired.

Closes #218

## How to test ##

This requires a full dev reset. The suggested way to do this:

```
docker-compose down
docker-compose build --no-cache
docker-compose up
(in other window) docker-compose exec app ./manage.py loaddata registrar/fixtures/*
```

Go to http://localhost:8080/edit_profile and edit the display name. Go to http://localhost:8080/admin/auditlog/logentry/ and verify that the change was recorded.

Go to http://localhost:8080/admin/registrar/domainapplication/add/ and add a domain application. Go to the detail view for the newly added application and click "History". Verify that you see the changes you've made.

![image](https://user-images.githubusercontent.com/15718530/201138382-7937bda0-8949-4ecb-ad45-34d62c679f2e.png)

![image](https://user-images.githubusercontent.com/15718530/201138580-46fc5616-5e63-4e41-84a6-f7c27bc9c5cc.png)
